nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/blob/DocumentBlobManagerComponent.java

```diff
@@ -361,22 +375,47 @@ protected String stripBlobKeyPrefix(String key) {
     // find which GCs to use
     // only GC the binary managers to which we dispatch blobs
     protected List<BinaryGarbageCollector> getGarbageCollectors() {
-        List<BinaryGarbageCollector> gcs = new LinkedList<>();
-        for (String providerId : getBlobDispatcher().getBlobProviderIds()) {
-            BlobProvider blobProvider = getBlobProvider(providerId);
-            BinaryGarbageCollector gc = blobProvider.getBinaryGarbageCollector();
-            if (gc != null) {
-                gcs.add(gc);
+        if (garbageCollectors == null) {
+            synchronized (this) {
+                if (garbageCollectors == null) {
+                    List<BinaryGarbageCollector> list = new ArrayList<>(); 
+                    for (String providerId : getBlobDispatcher().getBlobProviderIds()) {
+                        BlobProvider blobProvider = getBlobProvider(providerId);
+                        BinaryGarbageCollector gc = blobProvider.getBinaryGarbageCollector();
+                        if (gc != null) {
+                            list.add(gc);
+                        }
+                    }
+                    long idCount = list.stream().map(BinaryGarbageCollector::getId).distinct().count();
+                    sharedStorage = idCount < list.size();
+                    garbageCollectors = list;
+                }
             }
         }
-        return gcs;
+        return garbageCollectors;
+    }
+
+    /**
+     * Get the list of garbage collectors.
+     *
+     * @param refresh if true the list is recomputed, use latest computation otherwise
+     * @return a list of garbage collectors
+     * @since 10.10-HF56
+     */
+    protected List<BinaryGarbageCollector> getGarbageCollectors(boolean refresh) {
+        if (refresh) {
+            synchronized(this) {
+                garbageCollectors = null;
+            }
+        }
+        return getGarbageCollectors();
     }
 
     @Override
     public BinaryManagerStatus garbageCollectBinaries(boolean delete) {
         // do the GC in a long-running transaction to avoid timeouts
         return runInTransaction(() -> {
-            List<BinaryGarbageCollector> gcs = getGarbageCollectors();
+            List<BinaryGarbageCollector> gcs = getGarbageCollectors(true);
```

Comment:
We should explain in a comment why we need a "refreshed" list of GCs here and not elsewhere. Essentially, it's an optimization to avoid recomputing things too often while in the mark phase of a GC process.